### ISSUE TITLE
Excess electricity overestimated?

### ISSUE THREAD

**Issue by michieldenhaan** [quintel/etsource#1744]
The amount of excess electricity available seems to go up when using flexibility options.
I've made a scenario with lots of wind. When using no flex options, the amount of electricity curtailed or export is 3271 PJ. When using flex options, ``energy_power_hv_network_electricity`` decreases by 827 PJ (see screenshot). The demand of  ``energy_production_excess_electricity``, however, goes up by 873. The difference is 5.28%, which is equal to the HV network loss.

![energytransitionmodel httpetengine test 2018-06-18 15-23-02 1](https://user-images.githubusercontent.com/32056448/41538625-0b977146-730c-11e8-92d5-8de1b045d2b4.png)

![energytransitionmodel httpetengine test 2018-06-18 15-23-25 1](https://user-images.githubusercontent.com/32056448/41538669-1d196488-730c-11e8-8d6c-6eb3af766e7d.png)

Is it an option to add an edge between ``energy_power_hv_network_electricity`` and ``energy_production_excess_electricity``?

**Comment by michieldenhaan** [comment]
I think I figured out what's going on. See below for an example using no flex and using full flex:

**Scenario: NL 2016, 10.000 offshore wind turbines, no flex**

Total electricity production: **536.4 PJ**

Total used in sectors (transformer + own use): 340.16 PJ
Losses on electricity used in sectors: 18.96 PJ
Export/curtailment: 167.96 PJ
Losses on export/curtailment: 9.4 PJ
Total electricity demand: **536.4 PJ**


**Scenario: NL 2016, 10.000 offshore wind turbines, 12,279 power-to-hydrogen units**

Total electricity production: **536.4 PJ**

Total used in sectors (transformer + own use): 340.16 PJ
Losses on total used in sectors: 18.96 PJ
Extra losses reported in ETM: 9.4 PJ
Excess electricity used by flex: 177.32 PJ
Total electricity demand: **545.79 PJ**

Electricity demand (including losses) is larger than electricity production (+ 9.4 PJ), which shouldn't be possible. The reason is that total losses in the ETM are calculated based on production: 536.4 * 5.28% = 28.31 PJ (which equals ``losses on total used in sectors + extra losses reported in ETM``). Electricity available for flex, on the other hand, is calculated as: ``total electricity production - total used in sectors - losses on total used in sectors`` (= ``536.4 - 340.16 - 18.96= 177.3 PJ``). So losses on excess electricity are taken into account in the total loss calculation, but are not deducted in the total available excess electricity calculation. This results in the total demand exceeding total production (``excess electricity * loss = 177.3 * 5.28% = 9.4 PJ``).

I think we have two options here:
1. Deduct losses from the total available amount of excess electricity. That is, maximum available excess electricity demand is 177.32 PJ - 9.4 PJ = 167.9 PJ.
2. Do not take excess electricity into account when calculating network losses. I.e. lower total losses by 9.4 PJ.

I think I prefer option 2: one of the main reasons for using flex is that it can decrease the network load, so it seems a bit odd to take network losses into account (e.g. we currently calculate losses for curtailment..). On the other hand, it is possible in the ETM to use offshore wind for household batteries..

Including @jorisberkhout, @ChaelKruip

**Comment by antw** [comment]
> Scenario: NL 2016, 10.000 offshore wind turbines, 12,279 power-to-hydrogen units
> 
> Total electricity production: 536.4 PJ
> 
> Total used in sectors (transformer + own use): 340.16 PJ
> Losses on total used in sectors: 18.96 PJ
> Extra losses reported in ETM: 9.4 PJ
> Excess electricity used by flex: 177.32 PJ
> Total electricity demand: 545.79 PJ

How would I go about reproducing these numbers? Do they come from gqueries? As far as I can tell, supply and demand are perfectly balanced in Merit when there is excess wind and lots of H2 so I'd be curious to explore why this is not the case in the graph.

**Comment by michieldenhaan** [comment]
I calculated it manually by looking at ``energy_power_hv_network_electricity`` and ``energy_production_excess_electricity``. To split out losses, multiply a demand by ``loss factor / (1 - loss factor)``. So if total electricity used in sectors is 340.16 PJ, then the associated losses are ``340.16 PJ * 0.0528 / (1 - 0.0528)`` = ``340.16 * 0.0557`` = ``18.96`` PJ.
Total electricity required to be able to supply 340.16 PJ to the sectors is 340.16 (demand) + 18.96 (loss, 5.28% of required production) = 359.12 PJ.

**Comment by antw** [comment]
> Is it an option to add an edge between `energy_power_hv_network_electricity` and `energy_production_excess_electricity`?

It's not quite as simple as this, because not all excess electricity is high voltage. Therefore, not all excess would result in additional HV losses; even in the excess wind scenario, a simultaneous peak in wind *and* household solar PV would have excess HV and LV respectively.

It's likely we'd need separate LV, MV and HV versions of the `production_excess` node and even then I'm not sure if that will fix the apparent mismatch between the HV input and outputs, or where we would connect them (we could try it and find out).

> Do not take excess electricity into account when calculating network losses. I.e. lower total losses by 9.4 PJ.

This actually is already the case in the merit order; it considers only the "own use" and "sector use" of electricity when calculating HV loss. After Merit has run, ETEngine recalculates the loss based on the HV *inputs* and this is why there seems to be too much loss. I suspect the loss is correct, and that the issue is excess electricity "disappearing" from the HV node and reappearing in the excess_electricity node.

This is also a problem in production, so it's not related to the network calculation.

**Comment by michieldenhaan** [comment]
So if I understand this correctly, the following is happening:
- ETEngine assumes a fixed loss percentage for all centrally produced electricity, regardless of whether this electricity is used in a sector or whether it is curtailed/used for flex
- Merit only takes into account losses for centrally produced electricity that is used in a sector (or own use) when calculating available excess electricity. The difference between total production (HV, MV, LV) minus total demand (sector use + own use + associated losses) results in the amount of excess electricity. This is a gross amount, i.e. how much electricity is available without deducting any (possible) transport losses.
- Result: the same electricity appears at two different nodes in the graph: gross excess electricity ends up in `` energy_production_excess_electricity`` and losses for excess electricity produced by HV technologies end up in ``energy_power_hv_network_electricity``. This results in double counting: if we assume that there are no losses for flex/export/curtailment, available excess electricity should be equal to the gross amount and losses should be zero (option 2); if we assume that there are losses for flex/export/curtailment then available excess electricity should be equal to gross excess electricity minus losses (option 1). 

In more abstract terms:
``gross excess electricity`` = ``net excess electricity`` + ``transport losses``. Currently, the maximum of ``energy_production_excess_electricity`` is equal to ``gross excess electricity``, while ``energy_power_hv_network_electricity`` includes ``transport losses``. Hence, total energy in the graph is ``gross excess electricity`` + ``transport losses`` = ``net excess electricity`` + ``transport losses`` + ``transport losses`` -> double counting.

I see that option 1 can be quite complicated as not all excess electricity needs to come from HV technologies. But how about option 2:
> Do not take excess electricity into account when calculating network losses. I.e. lower total losses by 9.4 PJ.
Is it possible for ETEngine to only take into account losses for sector use + own use?

**Comment by ChaelKruip** [comment]
Removing milestone "Deploy June 2018" as this issue is already on production and it requires some thorough analysis and design/weighing of options. @michieldenhaan and I will return to this issue within two weeks time.

**Comment by michieldenhaan** [comment]
I think a related problem occurs for pumped storage:

**Scenario with lots of wind, no flex:**
![energytransitionmodel httpsbeta-engine energytransitionmodel com 2018-10-10 10-51-56](https://user-images.githubusercontent.com/32056448/46726215-2dbb5f00-cc7e-11e8-9216-a18d04c036ac.png)

**Same scenario with lots of wind, now with pumped storage**
![energytransitionmodel httpsbeta-engine energytransitionmodel com 2018-10-10 11-19-21](https://user-images.githubusercontent.com/32056448/46726304-5c393a00-cc7e-11e8-8afb-4d49c842889a.png)

I would expect that using pumped storage increases demand of ``energy_flexibility_pumped_storage_electricity`` and decreases ``energy_flexibility_curtailment_electricity `` and ``energy_export_electricity`` by exactly the same amount. What happens however is that the former increases by 293.38 PJ while the latter decrease by 280.33 PJ. As a result, total electricity demand goes up by 13.05 PJ. This is exactly equal to 293.38 * ``HV net loss`` (4.45%) (i.e. 293.38 * 0.00445 = 13.05). 

In other words, total electricity production has to be 13.05 PJ higher when using excess electricity for pumped storage than when curtailing/exporting this excess electricity. This should not be the case (or at least not in the above example in which we have an abundance of excess electricity).

I'm not entirely sure if the underlying cause of this is the same as described above. It might also be related to the fact that ``energy_flexibility_pumped_storage_electricity`` is connected to ``energy_power_hv_network_electricity`` (as opposed to other flex options like P2G or households P2P). When fixing the problem described in the previous posts we should check whether this problem is solved as well.

**Comment by antw** [comment]
Discussed with @michieldenhaan and @ChaelKruip, and we're considering this fixed by quintel/etengine@70d4209.

As @michieldenhaan described, the loss conversion of the HV network was a fixed percentage of the network node, but loss in Merit is based on HV-MV transformer demand + own use. The result is that the graph and Merit disagree on the amount of loss.

For example, see [production scenario 724491](https://engine.energytransitionmodel.com/inspect/724491/converters/energy_power_hv_network_electricity) which has maxed-out offshore wind.

The graph calculates 201.24B loss as this is 5.28% of all energy flowing through the node – which includes export and curtailment. However the loss given to Merit is only 18.95B, which is 5.28% of the HV-MV transformer, own use, and electricity flow caused by loss itself:

```ruby
PRODUCT(SUM(V(energy_power_hv_network_loss, electricity_input_curve)), 3600)
# => 18_953_810_000.0

V(energy_power_hv_network_loss, input_of_loss)
# => 201_241_524_746.4
```

The `input_of_loss` is far too high, because it includes export and curtailment. The extra electricity in the table comes from the graph over-calculating the amount of loss.

quintel/etengine@70d4209 changes ETEngine so that the *amount* of loss is fixed before the Merit calculation, and the conversion of HVloss←HVnet is adjusted after Merit runs to ensure the graph and Merit are consistent. This means that loss in a time-resolved graph does not include curtailment and export.

[Beta scenario 1270663](https://beta-engine.energytransitionmodel.com/inspect/1270663/converters/energy_power_hv_network_electricity):

```ruby
PRODUCT(SUM(V(energy_power_hv_network_loss, electricity_input_curve)), 3600)
# => 18_953_810_000.0

V(energy_power_hv_network_loss, input_of_loss)
# => 18_953_810_000.0
```

The graph and Merit both agree about the amount of loss, and the graph no longer causes unexpected loss to appear.

---

The result is that flexibility options *do not* count towards HV losses. This makes sense when the producer and flexibility technology are colocated (a household battery being charged by household PV, or wind-for-hydrogen turbines being used for P2G), but not when the electricity would flow through the network. However, excesses do not currently flow through the HV node, so I think this behaviour is correct for the graph as it is today.
